### DISCUSSION TITLE
Erlang 19.3 failed to load the crypto module after a Windows update

### DISCUSSION THREAD

**Discussion by MatiWQ** [rabbitmq/rabbitmq-server#8786] [Other] [answered]
Hi,

We are using an old version, 3.7.4 with Erlang 19.3. Last week RabbitMQ stopped working appareantly after windows updates. 
We try to install the same updates in a staing environment, but the issue didn't reproduce.

We tried to install latest OTP(20.3) version supported by 3.7.4 and it didn't work.
Only after finding that can be related to an OpenSSL bug or vulnerability and that it only affects 64 systems, I installed Erlang 20.3 32 bits and RabbitMQ started worked again.

Someone knows what can be the exact root cause of this issue and how can we reproduce it in another environment? 

There is another 64bit Erlang version we can install using the same RabbitMQ version that fixes the issue?

Thanks in advance.

Mati

This is the error in Event viewer:

Faulting application name: erl.exe, version: 0.0.0.0, time stamp: 0x58c85c74
Faulting module name: crypto.dll, version: 0.0.0.0, time stamp: 0x58c85f6b
Exception code: 0xc0000005
Fault offset: 0x0000000000017623
Faulting process id: 0x3af8
Faulting application start time: 0x01d9a9c66b055799
Faulting application path: C:\Program Files\Erlang\erts-8.3\bin\erl.exe
Faulting module path: C:\Program Files\Erlang\lib\crypto-3.7.3\priv\lib\crypto.dll
Report Id: 3d5651c3-feaa-41f2-9b29-e9b5e50c8f7c
Faulting package full name: 
Faulting package-relative application ID:`

**Comment by mkuratczyk** [comment]
RabbitMQ 3.7.4 is over 5 years old. Support for 3.7 ended almost 3 years
ago: https://www.rabbitmq.com/versions.html

Similar timelines apply to Erlang/OTP - you are using a version unsupported
for a few years. If you applied the operating system updates,
you now have a completely untested combination.

Best,

On Wed, Jul 5, 2023 at 6:10 PM MatiWQ ***@***.***> wrote:

> Hi,
>
> We are using an old version, 3.7.4 with Erlang 19.3. Last week RabbitMQ
> stopped working appareantly after windows updates.
> We try to install the same updates in a staing environment, but the issue
> didn't reproduce.
>
> We tried to install latest OTP(20.3) version supported by 3.7.4 and it
> didn't work.
> Only after finding that can be related to an OpenSSL bug or vulnerability
> and that it only affects 64 systems, I installed Erlang 20.3 32 bits and
> RabbitMQ started worked again.
>
> Someone knows what can be the exact root cause of this issue and how can
> we reproduce it in another environment?
>
> There is another 64bit Erlang version we can install using the same
> RabbitMQ version that fixes the issue?
>
> Thanks in advance.
>
> Mati
>
> This is the error in Event viewer:
>
> Faulting application name: erl.exe, version: 0.0.0.0, time stamp:
> 0x58c85c74
> Faulting module name: crypto.dll, version: 0.0.0.0, time stamp: 0x58c85f6b
> Exception code: 0xc0000005
> Fault offset: 0x0000000000017623
> Faulting process id: 0x3af8
> Faulting application start time: 0x01d9a9c66b055799
> Faulting application path: C:\Program Files\Erlang\erts-8.3\bin\erl.exe
> Faulting module path: C:\Program
> Files\Erlang\lib\crypto-3.7.3\priv\lib\crypto.dll
> Report Id: 3d5651c3-feaa-41f2-9b29-e9b5e50c8f7c
> Faulting package full name:
> Faulting package-relative application ID:`
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rabbitmq/rabbitmq-server/discussions/8786>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACI7PIV6E7OQN2YGDN5C5G3XOWGZDANCNFSM6AAAAAAZ7FSJFI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
Michał
RabbitMQ team

**Reply by MatiWQ** [reply]
Thanks Michal,

I know the support ended and that it's a really old version, but unfortunately we are running a .net legacy application with EasyNetQ 1.1.4.506. 
I am trying to understand what are our options. We need to continue using the legacy application, can we update RabbitMQ without changing EasyNetQ? 
There are any modifications we can do in the system(RabbitMQ, Erlang, Crypto library, etc) in order to overcome the issue?

Thanks,

Mati

**Reply by mkuratczyk** [reply]
I can't tell you how to overcome the issue. You are running unsupported software, as I understand combined with an operating system that didn't even exist when that software was supported. You are on your own when you do stuff like that.

As for the old EasyNetQ (or any other client application) - the compatibility between the client app and the server is in the protocol and that protocol doesn't change. You should obviously test your app with a new RabbitMQ version but I'd expect it to work.

**Reply by michaelklishin** [reply, marked as answer]
I'm quite certain that EasyNetQ 1.1 will be able to perform most of the key operations (if not all of them) against RabbitMQ 3.12 just fine.

RabbitMQ 3.12 supports .NET client going back all the way to 2.0 with very few behavior changes.

**Reply by MatiWQ** [reply]
Thanks for the help Michal and Michael.

Actually I was looking for a RabbitMQ and RabbitMQ.Client compatibility table but didn't found one.

EasyNetQ 1.1.4.506 is using RabbitMQ.Client 3.6.9.

I'll check the RabbitMQ upgrade path.

Thanks again.

**Comment by michaelklishin** [comment]
When it comes to OS updates, in particular Windows, they may affect the crypto library that ships with it.

Erlang 19.3 won't be compatible with whatever Windows might ship today. In fact, in the years between Erlang 19 and 25 or so, there was a renaissance of interest in TLS implementations, and many changes to OpenSSL, Erlang's TLS implementation, and likely the library that Windows ships with.
